src/pkg/artifact/controller.go

```diff
@@ -13,3 +13,103 @@
 // limitations under the License.
 
 package artifact
+
+import (
+	"context"
+	"github.com/goharbor/harbor/src/common/models"
+	"github.com/goharbor/harbor/src/pkg/artifact/manager"
+	"github.com/goharbor/harbor/src/pkg/artifact/model"
+	"time"
+)
+
+var (
+	// Ctl is a global Controller instance
+	Ctl = NewController()
+)
+
+// Controller is the only interface of artifact module to provide the
+// management functions for artifacts
+type Controller interface {
+	// List artifacts according to the query
+	List(context.Context, ...*model.Query) (total int64, artifacts []*model.Artifact, err error)
+	// Get the artifact specified by the ID
+	Get(context.Context, int64) (*model.Artifact, error)
+	// Create the artifact specified by the digest under the repository,
+	// returns the artifact ID and error. If tags are provided, attach them
+	// to the artifact.
+	Create(ctx context.Context, repository *models.RepoRecord, digest string, tags ...string) (id int64, err error)
+	// Delete just deletes the artifact record. The underlying data of registry will be
+	// removed during garbage collection
+	Delete(ctx context.Context, id int64) error
+	// GetSubResource returns the sub resource of the artifact
+	// The sub resource is different according to the artifact type:
+	// build history for image; values.yaml, readme and dependencies for chart, etc
+	GetSubResource(ctx context.Context, artifactID int64, resource string) ([]byte, error)
+	// Attach the tags to the artifact. If the tag doesn't exist, will create it.
+	// If the tag is already attached to another artifact, will detach it first
+	Attach(ctx context.Context, artifact *model.Artifact, tags ...string) error
+	// UpdateTag updates the tag
+	UpdateTag(ctx context.Context, tag *model.Tag) error
```

Comment:
what's the updateTag for?

Reply:
When pushing a same name but different digest image, we need to update the artifact_id field of the tag to point to the new image